<h1 id="validation-of-nonsynonymous-somatic-mutations-in-gallus-gallus-mareks-disease-md-lymphomas-via-agrixplex-targeted-dna-sequencing">Validation of Nonsynonymous Somatic Mutations in <em>Gallus gallus</em> Marek's Disease (MD) Lymphomas via Agrixplex Targeted DNA Sequencing</h1>
<p>We aim to validate somatic nonsynonymous snv and indel candidates (131 high priority candidates and 6 -- 50 bp regions of interest) called from whole genome sequencing in MD lymphomas in line 6x7 chickens infected with 1000 pfu of Marek's Disease Virus (MDV) JM102W at hatch. There were 188 samples tested including a cohort of gonadal tumors and matched normals collected from 2014; additional tumors collected in 2014 and 2017 from gonad, heart, spleen, thymus, pancreas, liver, proventiculus, kidney, and bursa; line 6 and line 7 (male and female) germline samples; and MD cell lines MSB1, RP2, and RP19. Targeted high coverage sequencing was performed on each variant and region of interest by Agriplex Genomics using the PlexSeq<sup>(TM)</sup> method and analyzed with PlexCall<sup>(TM)</sup> software. Whole genomic DNA was provided to Agriplex in 2 -- 96 well plates. These data represent raw reads from Agrixplex Genomics sequencing and the supporting documentation.</p>
<h2 id="candidate-variant-pipeline">Candidate Variant Pipeline</h2>
<p>Variants were originally called from 22 MD gonadal tumors and matched normals in the illustrated pipeline.</p>
<p><img src="./supporting_documentation/nonsynonymous_somatic_snv_indel_calling_pipeline.jpg" alt="drawing" width="1000px"/></p>
<p>Statistical modeling revealed the most important variable to predict true variants to be the number of callers agreeing on the variant. Variants called by 4 or more callers, whether snvs or indels, were the most likely to be validated.</p>
<h2 id="supporting-documentation">Supporting Documentation</h2>
<p>-General report of assay from Agriplex<br />
<code>supporting_documentation/AM-MSU_Cheng_Chicken_PlexSeq_Report_090517.docx</code></p>
<p>-Agriplex submission form<br />
<code>supporting_documentation/SNP_Submission_form_top150_w50bp.xlsx</code></p>
<p>-Candidate variants<br />
<code>supporting_documentation/somatic_snvs_and_indels_final_priorityv2.xlsx</code><br />
<code>supporting_documentation/somatic_snvs_and_indels_final_priorityv2.txt</code></p>
<p>-Sample annotation<br />
<code>supporting_documentation/MD_tumor_6x7_JM102W_annotation_database_2014_2017.xlsb</code></p>
<p>-Sample and plate organization<br />
<code>supporting_documentation/Sample_Name_Submission_Platemap_Agriplex_v3.xlsb</code></p>
<p>-Variant validation results from Agriplex<br />
<code>supporting_documentation/MSU_Cheng_Chicken_PlexSeq_Report_090517_SNPs_&amp;_Indels.xlsx</code><br />
<code>supporting_documentation/MSU_Cheng_Chicken_PlexSeq_Report_090517_Remaining_Indels.xlsx</code><br />
<code>supporting_documentation/MSU_Cheng_Chicken_PlexSeq_Report_090517_SNPs_Indels.txt</code></p>
<p>-Primer sequences flanking variants and regions of interest<br />
<code>supporting_documentation/100717_MSU_Chicken_PlexSeq_Primers.xlsx</code><br />
<code>supporting_documentation/100717_MSU_Chicken_PlexSeq_Primers.txt</code></p>
<h2 id="sequencing-files">Sequencing Files</h2>
<p>All raw FASTQ sequences are in <code>fastq_files</code></p>
<pre><code>$ find fastq_files -name &quot;*.fastq.gz&quot;
fastq_files/MSU_834-2-PL01-F03_S22_L001_R1_001.fastq.gz
fastq_files/MSU_834-0-PL01-E03_S21_L001_R1_001.fastq.gz
fastq_files/MSU_863-0-PL01-A05_S33_L001_R1_001.fastq.gz
fastq_files/MSU_822-1-PL02-A07_S143_L001_R1_001.fastq.gz</code></pre>
<h2 id="file-nomenclature">File Nomenclature</h2>
<p><code>MSU_101-T-PL01-C11_S83_L001_R1_001.fastq.gz</code><br />
<code>MSU_BirdID-Tissue-Plate#-Well#_Sample#_Lane#_Read#_AdditionalID.fastq.gz</code></p>
<p><code>MSU</code>: Michigan State University &quot;<strong>Go Green!</strong>&quot;<br />
<code>101</code>: Bird 101<br />
<code>T</code>: Thymus tumor<br />
<code>PL01</code>: Plate 1<br />
<code>C11</code>: Well C11<br />
<code>S83</code>: Sample 83<br />
<code>L001</code>: Lane 1<br />
<code>R1</code>: Read 1</p>
<h2 id="data-integrity">Data Integrity</h2>
<p>FastQC reports: <code>fastqc_raw</code></p>
<p>MD5 file path: <code>fastq_files/fastq_raw_reads_validation.md5</code></p>
<p>To perform MD5 check, from directory <code>fastq_files</code></p>
<pre><code>$ md5sum -c fastq_raw_reads_agrixplex_validation.md5
MSU_101-T-PL01-C11_S83_L001_R1_001.fastq.gz: OK
MSU_105-H-PL02-C02_S105_L001_R1_001.fastq.gz: OK
MSU_108-P-PL02-D02_S106_L001_R1_001.fastq.gz: OK
MSU_109-G-PL01-D09_S68_L001_R1_001.fastq.gz: OK</code></pre>
